LedgerHQ/ledger-live
# Bugfix/live 8698 Manager API v2 breaking changes adaptations

### 📝 Description

Some breaking changes have been introduced on the Manager API v2 (in staging so far).
Today we are using v1 of the API unless a feature flag `listAppsV2` is enabled, but we should still handle those breaking changes anyway because using the v2 API brings a big network optimisation and we should aim to activate the flag soon.

**Changes in the Manager v2 API (backend)**
- Optimisations
  - `/api/v2/apps/hash`: now outputs the apps in the same order as the input which makes it easier for us to use.
- Breaking changes:
  - `/api/v2/apps/hash`: now can return some apps as `null` if the hash does not match any app

**Changes in LL (this PR):**

- Both of those API changes are addressed in this PR.
- A new feature flag `listAppsV2minor1` is introduced in order to not activate the old one by mistake which would break old versions of LL.
- The polyfilling of app's data has also been fixed to match the behaviour of v1:
  - Previously, in list apps v2, the `currencyId` from the backend was never taken so in some cases ("Binance Chain" or "Persistence") there would be no associated currency, so in the My Ledger list of apps there would be a different icon, no ticker ("Persistence (XPRT)" with "(XPRT)" being the ticker), and no associated market cap, so the app would be at the bottom of the list when the "ranking by market cap" option is chosen.
  - Now the `currencyId` from the backend is taken by default and if it doesn't exist, it is polyfilled with the same logic as in list apps v1 (matching app's name and a `currency`'s `currency.name` or `currency.managerName`. 
- A fix regarding the uninstallation of apps which other installed apps depends on, from the `InstalledAppsModal`. For instance with Ethereum and Polygon installed, if you open the `InstalledAppsModal` listing all the installed apps, it would be impossible to uninstall Ethereum from there because the drawer `UninstallDependenciesModal` would be blocked by the currently opened modal (`InstalledAppsModal`).
  - The reason is that `InstalledAppsModal` was not using a `QueuedDrawer`, and also `UninstallDependenciesModal` should use the `isForcingToBeOpened` prop of `QueuedDrawer`. This is actually unrelated to listAppsV2 but it was noticed by QA so I fixed it here 🤷‍♂️ .
  - Refactor prop drilling nightmare of `setAppInstallWithDependencies`/`setAppUninstallWithDependencies` with a simple `React.Context`.
- Refactor `InstalledAppDependenciesModal` and `UninstallAppDependenciesModal` to have no business logic inside
- Rename action creator `installAppFirstTime` to `setHasInstalledAnyApp` for more clarity
- sorry i got a bit carried away with refactos unrelated to the initial purpose of the PR but it was itching me 😆 

**More context:** 
As a reminder this is the original work from Juan with more details about the optimisation in the PR description: https://github.com/LedgerHQ/ledger-live/pull/3097
And as an example of the optimisation, here's a before/after of networking when opening the manager on LLD:

before: <img width="320" alt="Screenshot 2023-10-02 at 14 22 12" src="https://github.com/LedgerHQ/ledger-live/assets/91890529/cf4e2a42-77ca-471f-80ad-4c926a3de5bd">
after: <img width="314" alt="Screenshot 2023-10-02 at 14 20 39" src="https://github.com/LedgerHQ/ledger-live/assets/91890529/7ca7990b-d32d-4bf2-ae3a-d0b3d407eb02">

**What I tested:**
- [x] LLD macOS
  - build from this PR
  - Infra VPN setup
  - open with `MANAGER_API_BASE=https://manager.api.live.stg.ledger-test.com/api /Applications/Ledger\ Live\ Beta.app/Contents/MacOS/Ledger\ Live\ Beta`
  - activate `listAppsV2minor1` feature flag
  - restart LLD
  - open manager
  - all apps installed on device are listed (eth, bitcoin, xrp, cardano ada, fido (this one is special because it doesn't have a fixed hash so the `/api/hash` returns null for it, which would break on a `develop` build), tron)
- [x] LLM Android
  - build from https://github.com/LedgerHQ/ledger-live/actions/runs/6394873605
  - Infra VPN setup
  - open app
  - activate `listAppsV2minor1` feature flag
  - set env var `MANAGER_API_BASE=https://manager.api.live.stg.ledger-test.com/api` in debug settings
  - open manager
  - all apps installed on device are listed (eth, bitcoin, xrp, cardano ada, fido (this one is special because it doesn't have a fixed hash so the `/api/hash` returns null for it, which would break on a `develop` build), tron)



### ❓ Context

- **Impacted projects**: `lld, llm, common, types-live` 
- **Linked resource(s)**: [LIVE-8698] 

### ✅ Checklist


### 📸 Demo



### 🚀 Expectations to reach

How to test ?
- [activate the Infra VPN](https://ledgerhq.atlassian.net/wiki/spaces/PKB/pages/3478650893/Infra+VPN+access+and+connect)
- the feature flag `listAppsV2minor1` must be enabled
- different APIs:
  - staging API (requires VPN): `MANAGER_API_BASE=https://manager.api.live.stg.ledger-test.com/api`
  - degraded staging API (required VPN) in case the other one returns 403 (no HTTPS): `MANAGER_API_BASE=http://appstore.api.aws.stg.ldg-tech.com/api`
    - LLD: also bypass CORS with `BYPASS_CORS=1`
    - LLM: does not work outside of debug builds because of HTTP cleartext
  - prod API (for now the changes are not in the prod api) `MANAGER_API_BASE=https://manager.api.live.ledger.com/api`
    - no additional config

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._




[LIVE-8698]: https://ledgerhq.atlassian.net/browse/LIVE-8698?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
